packages/abi-typegen/src/utils/parseTypeArguments.ts

```diff
@@ -0,0 +1,80 @@
+import type { IRawAbiTypeComponent } from '../interfaces/IRawAbiType';
+import type { IType } from '../interfaces/IType';
+import type { TargetEnum } from '../interfaces/TargetEnum';
+
+import { findType } from './findType';
+
+/*
+  Recursively parses the given `typeArguments` node
+*/
+export function parseTypeArguments(params: {
+  types: IType[];
+  target: TargetEnum;
+  typeArguments: IRawAbiTypeComponent[];
+  parentTypeId?: number;
+}): string {
+  const { types, typeArguments, parentTypeId, target } = params;
+
+  const attributeKey: 'inputLabel' | 'outputLabel' = `${target}Label`;
+
+  const buffer: string[] = [];
+
+  let parentType: IType | undefined;
+  let parentLabel: string | undefined;
+
+  if (parentTypeId !== undefined) {
+    parentType = findType({ types, typeId: parentTypeId });
+    parentLabel = parentType.attributes[attributeKey];
+  }
+
+  // loop through all `typeArgument` items
+  typeArguments.forEach((typeArgument) => {
+    const currentTypeId = typeArgument.type;
+    const currentType = findType({ types, typeId: currentTypeId });
+    const currentLabel = currentType.attributes[attributeKey];
+
+    if (typeArgument.typeArguments) {
+      // recursively process nested `typeArguments`
+      const nestedParsed = parseTypeArguments({
+        types,
+        target,
+        parentTypeId: typeArgument.type,
+        typeArguments: typeArgument.typeArguments,
+      });
+
+      buffer.push(nestedParsed);
+    } else {
+      buffer.push(`${currentLabel}`);
+
+      /*
```

Comment:
These comments are some kind of TODO?

Reply:
No, just some annotations I don't want to risk losing just yet.

Planning on doing so right before (or _after_) the merge.

Reply:
maybe it should be removed now as the PR is ready to review/merge...